// angular2-http-interceptor/src/http_interceptor_module.js
import { NgModule } from '@angular/core';
import { Http, RequestOptions, XHRBackend, BaseRequestOptions, ResponseOptions, BaseResponseOptions, XSRFStrategy, CookieXSRFStrategy, BrowserXhr } from '@angular/http';
import { HttpInterceptor } from './http_interceptor';
import { HttpInterceptorBackend } from './http_interceptor_backend';
/**
 * @param {?} httpInterceptorBackend
 * @param {?} requestOptions
 * @return {?}
 */
export function httpFactory(httpInterceptorBackend, requestOptions) {
    return new Http(httpInterceptorBackend, requestOptions);
}
/**
 * @return {?}
 */
export function createDefaultCookieXSRFStrategy() {
    return new CookieXSRFStrategy();
}
/**
 * The module that includes http interceptor providers
 */
export var HttpInterceptorModule = (function () {
    function HttpInterceptorModule() {
    }
    /**
     * @param {?} interceptors
     * @return {?}
     */
    HttpInterceptorModule.withInterceptors = function (interceptors) {
        return {
            ngModule: HttpInterceptorModule,
            providers: interceptors.concat([
                {
                    deps: [HttpInterceptor, XHRBackend],
                    provide: HttpInterceptorBackend,
                    useClass: HttpInterceptorBackend,
                },
                {
                    deps: [HttpInterceptorBackend, RequestOptions],
                    provide: Http,
                    useFactory: httpFactory,
                },
                BrowserXhr,
                {
                    provide: RequestOptions,
                    useClass: BaseRequestOptions
                },
                {
                    provide: ResponseOptions,
                    useClass: BaseResponseOptions
                },
                XHRBackend,
                {
                    provide: XSRFStrategy,
                    useFactory: createDefaultCookieXSRFStrategy
                },
            ])
        };
    };
    HttpInterceptorModule.decorators = [
        { type: NgModule },
    ];
    /** @nocollapse */
    HttpInterceptorModule.ctorParameters = function () { return []; };
    return HttpInterceptorModule;
}());
function HttpInterceptorModule_tsickle_Closure_declarations() {
    /** @type {?} */
    HttpInterceptorModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    HttpInterceptorModule.ctorParameters;
}
